### DISCUSSION TITLE
Storage template do not reflect the settings set and use UUID folder structure

### DISCUSSION THREAD

**Discussion by SteezCram** [immich-app/immich#8389] [Q&A]
### The bug

When using the storage template and the hash verification the folder structure set in the storage template is not taken into account. Instead, the folder structure is UUID-based. I try to run the storage template migration but it does not do anything.
I have attached some pictures of my problem.

Here is my settings:
![image](https://github.com/immich-app/immich/assets/46353105/5122ec5b-a490-4025-b29e-85a60b5d71ad)

This is inside my profile folder (that uses my profile ID):
![image](https://github.com/immich-app/immich/assets/46353105/cc7974ed-0b4f-4a37-8807-30616d307fde)

Inside the folder `00`:
![image](https://github.com/immich-app/immich/assets/46353105/5ea00317-1752-4139-b7d0-e0f8af9cdb11)

Inside the folder `3c`:
![image](https://github.com/immich-app/immich/assets/46353105/257fa629-ce7c-4aab-9a40-d37179b971f9)

This is the same folder structure across all my files with only one file per folder (but since it is a UUID it makes sense, I have only ~3000 files).

### The OS that Immich Server is running on

Docker

### Version of Immich Server

v1.100.0

### Version of Immich Mobile App

v1.100.0

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
default settings
```

### Your .env content

```Shell
default settings
```

### Reproduction steps

```bash
1. Use the mobile application to upload some pictures and videos
2. Use storage template option with hash verification enabled
3. Click on the storage migration button
```

**Comment by alextran1502** [comment]
Are you checking the `library` folder or the `upload` folder? storage template will move file into the `library` folder

**Comment by SteezCram** [comment]
Both the `library` and the `uploads` folders are configured to the same folder: `/mnt/data1/immich`. In addition, the `video` folder should go to this folder too. Only the `thumbs` folder is set to a persistent volume claim.

Maybe I misconfigured something on my side.

**Comment by bo0tzz** [comment]
> Both the library and the uploads folders are configured to the same folder: /mnt/data1/immich.

That doesn't sound right. You should probably just use the default mounts from the docker-compose setup we supply.

**Comment by SteezCram** [comment]
> That doesn't sound right. You should probably just use the default mounts from the docker-compose setup we supply.

I use Truenas Scale and if I use the PVC I cannot access them easily outside the container. This is why I set this like that.

But I did notice something weird where both of the data are in my PVC and on my disk. I think I did something wrong and I will try to set it again following the folder structure you mentioned.

**Comment by SteezCram** [comment]
I have investigated a bit more and I came across these errors for all the files:
```json
ERROR [StorageTemplateService] Problem applying storage template
ERROR [StorageTemplateService]  Error: EPERM: operation not permitted, copyfile 'upload/upload/beff4a59-92c7-4aff-b450-32d126e5c155/c5/3b/c53b0510-aff8-4062-bcb8-58bba026bb65.mp4' -> 'upload/library/thomas/2024/02/25/VID20240225013642.mp4'
ERROR [StorageTemplateService] Object:
{
"id": "976e81f3-994b-41b3-9d55-2ef971333e5d",
"oldPath": "upload/upload/beff4a59-92c7-4aff-b450-32d126e5c155/c5/3b/c53b0510-aff8-4062-bcb8-58bba026bb65.mp4",
"newPath": "upload/library/thomas/2024/02/25/VID20240225013642.mp4"
}
```

So it does not take into account the custom path actually and sticks with the one by default. I think this was not the behavior in the previous version because, with the same config, I did my files were in the right location.

**Reply by mmomjian** [reply]
This error (`EPERM: operation not permitted`) indicates that there is an issue with the permissions on your filesystem and Immich is not able to perform the requested move. You will need to debug your mounts and folder/subfolder permissions - Immich has no control over that.

**Reply by leocb** [reply]
Hi @mmomjian, I have the same problem. Can you tell us what are the credentials being used when copying the files? I ask because if I bash into the microservices container, I can do these operations normally with the root user, so I don't know what permission I'm doing wrong here.

In my case, The correct folder structure is created, but no files are copied.

**Reply by leocb** [reply]
Found the solution in [this reddit comment](https://www.reddit.com/r/truenas/comments/1b1weo9/comment/l132fuj/), seems like it's an issue with js using chmod wrongly(?)
In my case, the library folder is in a TrueNAS Scale server. setting the ACL Mode to Passthrough solved this.

In the TrueNAS web UI, go to Datasets -> Select your data set -> Dataset Details [EDIT] -> Advanced Options -> ACL Mode -> Set to `Passthrough`

**Reply by MatrixTechno** [reply]
> Found the solution in [this reddit comment](https://www.reddit.com/r/truenas/comments/1b1weo9/comment/l132fuj/), seems like it's an issue with js using chmod wrongly(?) In my case, the library folder is in a TrueNAS Scale server. setting the ACL Mode to Passthrough solved this.
> 
> In the TrueNAS web UI, go to Datasets -> Select your data set -> Dataset Details [EDIT] -> Advanced Options -> ACL Mode -> Set to `Passthrough`

Thank you so much, I set this in my 'library' dataset and it immediately fixed the issue. I've been breaking my head against a wall for hours. Thanks so much.

**Comment by SteezCram** [comment]
Any plan to add/fix this behavior?

**Comment by pr0927** [comment]
This must be what's happening for me - I am also on TrueNAS SCALE, using the TrueCharts version of Immich.

Uploaded files don't seem to actually show up in the folder designated as uploads, and no amount of time or job runs seems to make anything move to the folder designated as library.

I have both mapped as NFS paths (internally to folders designated as dual NFS/SMB shares) instead of PVC stuff, which is largely inaccessible on TrueNAS SCALE.
